### ISSUE TITLE
oc login should avoid clobbering kubeconfig contexts

### ISSUE THREAD

**Issue by TheLonelyGhost** [openshift/oc#283]
Default behavior of `oc login https://openshift.example.com --username=thelonelyghost --password='hunter2'` is to do the following:

- Setup a new (or if exists, update) cluster in the kubeconfig by the name of `openshift-example-com:443`
- Setup a new (or if exists, update) user in the kubeconfig by the name of `thelonelyghost/openshift-example-com:443`
- Setup a new (or if exists, update) context in the kubeconfig by the name of `default/openshift-example-com:443/thelonelyghost`.
- Set the current context to be `default/openshift-example-com:443/thelonelyghost`

This is fine for initial setup, but periodically I'm forced to login again to refresh the bearer token stored in the user section of the kubeconfig. When I do that, I must run the same `oc login` command again. Here's the problem:

If I re-run that command, it clobbers any changes I've made to human-facing names of the cluster, the context, or the user that have zero functional effect on the file. If I renamed the `default/openshift-example-com:443/thelonelyghost` context to be more user-friendly, such as `company-prod`, logging in again adds an additional context of `default/openshift-example-com:443/thelonelyghost` back again. It does not update credentials in-place, it adds them if they're not named exactly what is expected.

I've written a tool in python named [`oc-replacement`](https://github.com/thelonelyghost/oc-replacement) to re-authenticate in-place as I desire, but it would be much more desirable for this to be default behavior or even an option with the `oc` command line tool.

**Comment by OddKMS** [comment]
I'm hoping that this gets a bit of attention - it is quite annoying to have to deal with.

In addition, this completely mangles the `kube-ps1` addon as it uses the separate config fields (context name, user name, cluster name, etc) to build up its output. 

Using a normal kubeconfig it would look like this:
`(⎈ |contextname:namespace)`

Instead when using `oc login` the result is this:
`(⎈ |namespacename/clusterservername:port/longformusername:namespace)`

On the whole it is not a serious issue, but when have we devs ever let that stop us from complaining about something?

**Comment by reitzig** [comment]
What's curious is that `certificate-authority` is taken from an existing cluster with matching server, but then a new cluster is created anyway. So, there's _some_ reading of the config file going on already.

My main pain point is indeed the long cluster names and longer context names that bloat (and kill) my [shell prompt](https://github.com/starship/starship) (in smaller windows).

**Comment by flippyboy** [comment]
I have created a support case with Red Hat about this and got the following reply:
**After careful analysis, we've determined that resolving this problem requires a significant effort which we're not able to allocate to this issue at the moment.**

If others with active support agreements feel that this is important to them, you can reference [this](https://access.redhat.com/solutions/7079565) when creating your own support request. If more people create requests, maybe it will get enough attention.

**Comment by adietish** [comment]
This bug unfortunately hits us in vscode-openshift especially when trying to support multiple configs listed in the KUBECONFIG env variable: https://github.com/redhat-developer/vscode-openshift-tools/issues/4816

Is there any update on where this issue and the support are going, please?
@vrubezhny can you please provide more details?

**Comment by vrubezhny** [comment]
Actually, there's not much to add here - everything has already been said in the comments to the issue. That's a real problem. 

The only solution I found is using `--kubeconfig` argument to redirect any Kube configuration changes on Login operation to a temporary Kube config file and then manually merging the changed objects back to the according Kube configs that are configured on `$KUBECONFIG` env,. variable.

**Comment by oscrx** [comment]
This tool might help some people here: https://github.com/corneliusweig/konfig

**Comment by vkrizan** [comment]
Could the `oc login` use the `--context` option value to store the login credentials? Currently the `--context` option is ignored by this command.

**Comment by pietergobin** [comment]
So I provided a quick fix to this and have opened a PR :) Hopefully it helps some people. I admit I just put it behind a --no-clobber flag for now but am open to feedback on how to better resolve this in the future 

https://github.com/openshift/oc/pull/2185

**Comment by adietish** [comment]
Here are the steps that showcase how `oc login` adds all contexts, clusters, users that I have in config files (listed in `KUBECONFIG`) to my main config.
This is very annoying because the new entries, that it creates in my main config are shadowing the ones in my other files (listed in my `KUBECONFIG`). This is because when merging kube configs, the first entry "wins".

**Steps:**
1. ASSERT:
```bash
$ echo $KUBECONFIG
/Users/adietish/.kube/config:/Users/adietish/.kube/sandbox.yaml
```
```bash
$ cat ~/.kube/config
apiVersion: v1
kind: Config
current-context: sandbox
preferences: {}
```
```bash
$ cat sandbox.yaml
apiVersion: v1
kind: Config
current-context: sandbox
preferences: {}
clusters:
- cluster:
    insecure-skip-tls-verify: true
    server: https://api.XXXX.openshiftapps.com:6443
  name: sandbox
contexts:
- context:
    cluster: sandbox
    namespace: adietish-dev
    user: sandbox
  name: sandbox
users:
- name: sandbox
  user:
    token: sha256~XXXXX
```
2. EXEC: `oc login --web`

**Result:**
The user and cluster that exists in `~/.kube/sandbox.yaml` were added to `~/.kube/config` which only had `current-context`.
```bash
$ cat ~/.kube/config
apiVersion: v1
clusters:
- cluster:
    server: https://api.XXXX.openshiftapps.com:6443
  name: api-XXXX-openshiftapps-com:6443
contexts:
- context:
    cluster: api-rm1-0a51-p1-openshiftapps-com:6443
    namespace: adietish-dev
    user: adietish/api-XXXX-openshiftapps-com:6443
  name: adietish-dev/api-XXXX-openshiftapps-com:6443/adietish
- context:
    cluster: sandbox
    namespace: adietish-dev
    user: sandbox
  name: sandbox
current-context: adietish-dev/api-XXXX-openshiftapps-com:6443/adietish
kind: Config
users:
- name: adietish/api-XXXX-openshiftapps-com:6443
  user:
    token: sha256~XXXX-9E
```

**Comment by tchap** [comment]
@adietish This seems to be a different issue, right?

**Comment by adietish** [comment]
@tchap: I'd think that oc creating a new context and user is what the user of this issue reported initially. Then the addition existing clusters, users, etc into the main config is afaics a different/new issue. Want me to create a new one?

**Comment by TheLonelyGhost** [comment]
The original issue was, indeed, scoped to adding a new context when it didn't need to. In the POC written in python, I also added support for multiple files in KUBECONFIG since that appeared to be a supported standard. In retrospect, some tools did not consistently honor KUBECONFIG having multiple files in a PATH-like structure and I abandoned that idea not long after.

So yes it is a concern I share, but out of scope for the issue I was trying to raise here. 😄

**Comment by tchap** [comment]
@adietish Yes please, create a separate issue. Best to use the company Jira, if you can. Thanks.
